File: python-sdk/demo/harvester/world.py
import logging

from typing import Dict, Iterable
from expansion import types


class World:
    def __init__(self):
        self._logger = logging.getLogger("World")
        self.asteroids: Dict[int, types.PhysicalObject] = {}

    def update_object(self, obj: types.PhysicalObject):
        if obj.object_type == types.ObjectType.ASTEROID:
            asteroid = obj
            cached_asteroid = self.asteroids.setdefault(
                asteroid.object_id,
                types.PhysicalObject(object_id=asteroid.object_id,
                                     object_type=asteroid.object_type,
                                     position=asteroid.position,
                                     radius=asteroid.radius))
            cached_asteroid.update(asteroid)

    def update_objects(self, objects: Iterable[types.PhysicalObject]):
        for obj in objects:
            self.update_object(obj)

File: typescript-sdk/demo/harvester/world.ts
import type { PhysicalObject } from "../../types/index.js";

export class World {
    readonly asteroids = new Map<number, PhysicalObject>();

    update_object(object: PhysicalObject): void {
        if (object.object_type !== "asteroid") {
            return;
        }
        const cached = this.asteroids.get(object.object_id);
        if (!cached || object.position.timestamp >= cached.position.timestamp) {
            this.asteroids.set(object.object_id, {
                object_type: object.object_type,
                object_id: object.object_id,
                position: object.position,
                radius: object.radius,
            });
        }
    }

    update_objects(objects: Iterable<PhysicalObject>): void {
        for (const object of objects) {
            this.update_object(object);
        }
    }
}
